rrfs-test/CMakeLists.txt

```diff
@@ -73,25 +77,12 @@ if (CLONE_RRFSDATA)
          set(MPI_ARGS "${MPI_ARGS} --exclusive")
       endif()   
 
-      # JEDI test cases
-      # ----------------
-      # NOTE: To add a new ctest for MPAS, simple add the new yaml file here (without the extension)
-      #       Also need to add the associated executable to the rrfs_fv3jedi_exes list below 
-      list( APPEND rrfs_mpasjedi_tests
-             rrfs_mpasjedi_2022052619_Ens3Dvar
-             rrfs_mpasjedi_2022052619_letkf
-             rrfs_mpasjedi_2022052619_atms_npp_qc
-          )
-
-      # JEDI executables for each test case above
-      list ( APPEND rrfs_mpasjedi_exes
-             mpasjedi_variational.x
-             mpasjedi_enkf.x
-             mpasjedi_variational.x
-      )
-
-      foreach(case IN LISTS rrfs_mpasjedi_tests)
-
+      foreach(keypair ${rrfs_mpasjedi_tests})
```

Comment:
Should we use `get_from_dictionary `here?

Reply:
We could, but the `get_from_dictionary` function requires you to pass in the key name (test case name in this case). And in order to get the key name from the dict that is being looped over, I needed to use the `string(REGEX MATCH "([^=]+)=([^=]*)" _ ${keypair})` line. This command returns both the key and the value as `${CMAKE_MATCH_1}` and `${CMAKE_MATCH_2}` so I just went ahead and used it instead of calling the `get_from_dictionary` function. But I left that function in `funcs.cmake` for completeness and in case we need it later.

Reply:
I think we could try a little different use of this dictionary as I would add in the comment with a label #dic1

Reply:
So just to confirm @TingLei-NOAA, you are recommending replacing: 

`set(exe ${CMAKE_MATCH_2})`

with

`get_from_dictionary(rrfs_fv3jedi_tests case exe)` 

I supposed that is good since it shows how to use the new get_from_dictionary function. There is no functional difference in the result but either way works for me.

Reply:
Sam,   Basically, I suggest we don't need to use " foreach(keypair ${rrfs_mpasjedi_tests})"  with rrfs_mpasjedi/fv3jedi_tests being converted to dic-like structure. 
The added dic-like structure could be like  rrfs_tests_execs as I used in #dic1 comment.  we just need set it up with the key is the ctest name and the value is the corresponding executable.   By doing this, we don't need extra processing as the line 41, 42. I feel this way has a better readability by using the dic-like structure introduced in your PR.

Reply:
Sorry Ting, I don't think I understand. How do we iterate through the dict if we do not use `foreach(keypair ${rrfs_mpasjedi_tests})`? And isn't the code I added to `CMakeLists.txt` setting up the dict just as you describe where the key is the ctest name and the value is the executable? 

```
set(rrfs_fv3jedi_tests)
add_to_dictionary(rrfs_fv3jedi_tests "rrfs_fv3jedi_hyb_2022052619"     "fv3jedi_var.x")
add_to_dictionary(rrfs_fv3jedi_tests "rrfs_fv3jedi_letkf_2022052619"   "fv3jedi_letkf.x")
```

Reply:
Sam, one thing is that in current version, the rrfs_fv3/mpas_jedi_tests  are converted to dic-like structures, which make the following line 80 is needed since the former are not a list again.  In my opinion, by using a new dic-like variable would make the changes simpler.     I could make an example on top of your codes as soon as possible.

Reply:
Yes maybe it would be better for you to show me an example so I understand a little better.

Reply:
@SamuelDegelia-NOAA I created a draft PR to show what I referred to. https://github.com/NOAA-EMC/RDASApp/pull/82.    Hopefully, this is simpler/easier to understand.  
You could have a different version following your original plan by let the dict-like rrfs_fv3jedi_test to have more methods to iterate over elements .  The key is that 1), it is better we don't need extra pattern processing (on pair) in the CMakefile.txt, it is better to be hidden in the function level, 2) the structure like rrfs_fv3jedi_tests is better to be able to provide a method to give a list of ctest.   That is my two cents.  Thanks.

Reply:
@TingLei-NOAA so in your version you use a combination of lists and dicts? I see you loop through rrfs_fv3jedi_tests (containing the ctest names), and then use the ctest name as the key to retrieve the value (exe name) in the dict? If someone were adding a new ctest, wouldn't this mean that they need to add the new test both to the list and to the dict? 

```
     list( APPEND rrfs_fv3jedi_tests
             rrfs_fv3jedi_hyb_2022052619
             rrfs_fv3jedi_letkf_2022052619
             NEW_CTEST_NAME
          )
      add_to_dictionary(rrfs_test_execs "rrfs_fv3jedi_hyb_2022052619"     "fv3jedi_var.x")
      add_to_dictionary(rrfs_test_execs "rrfs_fv3jedi_letkf_2022052619"   "fv3jedi_letkf.x")
      add_to_dictionary(rrfs_test_execs "NEW_CTEST_NAME"                    "NEW_EXE.x")
```

I thought the goal of this PR was to be more concise on adding a new CTest so that you wouldn't need to add to both the test name lists and the exe list. In my version, all you need to do to add a new ctest is add one line at the top of the file: 

`add_to_dictionary(rrfs_fv3jedi_tests "NEW_CTEST_NAME"     "NEW_EXE.x")`

I understand not wanting to include the `string(REGEX MATCH "([^=]+)=([^=]*)" _ ${keypair})` and instead keep that at the function level, but it still seems simpler from my perspective.  If this is the route we want to go though, I am okay with it.

Reply:
@SamuelDegelia-NOAA  Yes, in my version,  the developer need to add their test to the test lists first.  Normally, I prefer readability the first.  As I said, your method with only one dic-like structure would be more advantageous , from my point of view, if  an iterator method could be added to your dic-like structure data  while the pattern processing on "pair" is not needed.  Basically, such a pattern processing over a pair from the dic make some loss of the expected readability and simplicity introducing the dic-like structure (when the pattern processing is difficult to understand)

Reply:
Okay, so the primary issue then is with the REGEX statement. I can see if I can move that into `funcs.cmake` somehow.

Reply:
Sam, Yes,  with your strategy, only a method to get all the keys ( rather than we have to get those pair and do processing again) would be enough .   From the chatgpt:).  Such a function could be like 
```
function(get_all_keys dict result)
    set(keys "")
    foreach(entry IN LISTS ${dict})
        string(FIND "${entry}" "=" EQUAL_POS)
        if(EQUAL_POS GREATER -1)
            string(SUBSTRING "${entry}" 0 ${EQUAL_POS} ENTRY_KEY)
            list(APPEND keys "${ENTRY_KEY}")
        endif()
    endforeach()
    set(${result} "${keys}" PARENT_SCOPE)
endfunction()
```.
Now, i think the most perferable way is your current one with an enhancement similar to the above one.

Reply:
Okay I'll work on adding that in on Monday. I agree with @guoqing-noaa that we are likely over-engineering this problem, so hopefully things will be good to go after this change.